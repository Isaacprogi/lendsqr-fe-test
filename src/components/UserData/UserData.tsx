import { useNavigate } from 'react-router-dom'
import DotsVertical from '../../assets/svg/user-activities/dots-vertical.svg'

import ActivateUser from '../../assets/svg/user-activities/ActivateUser.svg'
import BlacklistUser from '../../assets/svg/user-activities/BlacklistUser.svg'
import ViewDetails from '../../assets/svg/user-activities/ViewDetails.svg'
import { userDataProps } from '../../utils/types/interface_and_types'


export const UserData = ({ user,displayStatusOptions,setDisplayStatusOptions }: userDataProps) => {
  const navigate = useNavigate()

  const handleActivateUser = () => {
    user.status = 'Active'
    setDisplayStatusOptions('')
  }
  const handleBlackListUser = () => {
    user.status = 'Blacklisted'
    setDisplayStatusOptions('')
  }
  const handleViewDetails = () => {
    navigate(`/user-details/${user.id}`)
    setDisplayStatusOptions('')
  }
   const handleDisplay = (Data:{id:string}) => {
      if(displayStatusOptions){
        return setDisplayStatusOptions('')
      }
      setDisplayStatusOptions(Data.id)
   }

  return (
    <div data-testid='user-data-box' className="table-data">
      <span className='organization'>{user.orgName}</span>
      <span className='username'>{user.userName}</span>
      <span className='email'>{user.email}</span>
      <span className='phone-number'>{user.phoneNumber}</span>
      <span className='date-joined'>{new Date(user?.createdAt).toDateString()}</span>

      <span className='status-span'>
        <div data-testid="status-options-id" className={`status-options ${user?.id === displayStatusOptions ? "visible" : ""} `}>
          <span data-testid="view-list-span" onClick={handleViewDetails}>
            <img src={ViewDetails} alt="" />
            View Details
          </span>
          <span data-testid="black-list-span" onClick={handleBlackListUser}>
            <img src={BlacklistUser} alt="" />
            Blacklist User
          </span>
          <span data-testid="active-list-span" onClick={handleActivateUser}>
            <img src={ActivateUser} alt="" />
            Activate User
          </span>
        </div>
        <div className={`pending ${user?.status}`}>{user?.status}</div>
        <img onClick={()=>handleDisplay({id:user?.id})} src={DotsVertical} alt="three-dots" />
      </span>
    </div>
  )
}
