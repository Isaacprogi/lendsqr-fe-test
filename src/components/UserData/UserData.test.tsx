import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import { UserData } from './UserData'
import { BrowserRouter } from 'react-router-dom';
import { useState } from 'react';
import renderer from 'react-test-renderer'
const dataProps = {
    user: {
        id: '1',
        orgName: 'lendsqr',
        userName: 'rema',
        email: "rema@gmail.com",
        phoneNumber: "09157065783",
        createdAt: "ririir",
        status:'Active'
    }
}

describe('test for NavBar component', () => {

    const UserDataComponent = () => {
        const [displayStatusOptions, setDisplayStatusOptions] = useState<string>('')
        return <UserData user={dataProps.user} displayStatusOptions={displayStatusOptions} setDisplayStatusOptions={setDisplayStatusOptions}/>
    }

    test('props shoud be present', () => {
        render(<BrowserRouter>
            <UserDataComponent/>
             </BrowserRouter>)
        expect(screen.getByText('lendsqr')).toBeInTheDocument()
        expect(screen.getByText('rema')).toBeInTheDocument()
        expect(screen.getByText('rema@gmail.com')).toBeInTheDocument()
        expect(screen.getByText('09157065783')).toBeInTheDocument()       
    })

    test('status options should display properly', async()=>{
        render(<BrowserRouter>
            <UserDataComponent/>
             </BrowserRouter>)
        waitFor(()=>{
            const buttonElement = screen.getByRole('img', {name:"three-dots"})
            const divELement = screen.getByTestId('status-options-id')
            const viewElement = screen.getByTestId('view-list-span')
            const blackElement = screen.getByTestId('black-list-span')
            const activeElement = screen.getByTestId('active-list-span')
            expect(divELement).not.toHaveClass("visible")
            fireEvent.click(buttonElement)
            expect(divELement).toHaveClass('visible')
            fireEvent.click(buttonElement)
            expect(divELement).not.toHaveClass('visible')
            fireEvent.click(buttonElement)
            expect(divELement).toHaveClass('visible')

            fireEvent.click(viewElement)
            expect(divELement).not.toHaveClass("visible")

            fireEvent.click(blackElement)
            expect(divELement).not.toHaveClass("visible")
            expect(dataProps.user.status).toBe('Blacklisted')

            fireEvent.click(activeElement)
            expect(divELement).not.toHaveClass("visible")
            expect(dataProps.user.status).toBe('Active')
        })
        

    })

    test('should match snapshot', async () => {
        const tree = renderer.create(<BrowserRouter>
            <UserDataComponent/>
             </BrowserRouter>).toJSON();
        expect(tree).toMatchSnapshot();
      });
})