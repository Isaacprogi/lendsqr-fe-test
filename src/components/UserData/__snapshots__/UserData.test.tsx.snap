// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`test for NavBar component > should match snapshot 1`] = `
<div
  className="table-data"
  data-testid="user-data-box"
>
  <span
    className="organization"
  >
    lendsqr
  </span>
  <span
    className="username"
  >
    rema
  </span>
  <span
    className="email"
  >
    rema@gmail.com
  </span>
  <span
    className="phone-number"
  >
    09157065783
  </span>
  <span
    className="date-joined"
  >
    Invalid Date
  </span>
  <span
    className="status-span"
  >
    <div
      className="status-options  "
      data-testid="status-options-id"
    >
      <span
        data-testid="view-list-span"
        onClick={[Function]}
      >
        <img
          alt=""
          src="/src/assets/svg/user-activities/ViewDetails.svg"
        />
        View Details
      </span>
      <span
        data-testid="black-list-span"
        onClick={[Function]}
      >
        <img
          alt=""
          src="/src/assets/svg/user-activities/BlacklistUser.svg"
        />
        Blacklist User
      </span>
      <span
        data-testid="active-list-span"
        onClick={[Function]}
      >
        <img
          alt=""
          src="/src/assets/svg/user-activities/ActivateUser.svg"
        />
        Activate User
      </span>
    </div>
    <div
      className="pending Active"
    >
      Active
    </div>
    <img
      alt="three-dots"
      onClick={[Function]}
      src="/src/assets/svg/user-activities/dots-vertical.svg"
    />
  </span>
</div>
`;

exports[`test for NavBar component should match snapshot 1`] = `
<div
  className="table-data"
  data-testid="user-data-box"
>
  <span
    className="organization"
  >
    lendsqr
  </span>
  <span
    className="username"
  >
    rema
  </span>
  <span
    className="email"
  >
    rema@gmail.com
  </span>
  <span
    className="phone-number"
  >
    09157065783
  </span>
  <span
    className="date-joined"
  >
    Invalid Date
  </span>
  <span
    className="status-span"
  >
    <div
      className="status-options  "
      data-testid="status-options-id"
    >
      <span
        data-testid="view-list-span"
        onClick={[Function]}
      >
        <img
          alt=""
          src="ViewDetails.svg"
        />
        View Details
      </span>
      <span
        data-testid="black-list-span"
        onClick={[Function]}
      >
        <img
          alt=""
          src="BlacklistUser.svg"
        />
        Blacklist User
      </span>
      <span
        data-testid="active-list-span"
        onClick={[Function]}
      >
        <img
          alt=""
          src="ActivateUser.svg"
        />
        Activate User
      </span>
    </div>
    <div
      className="pending Active"
    >
      Active
    </div>
    <img
      alt="three-dots"
      onClick={[Function]}
      src="dots-vertical.svg"
    />
  </span>
</div>
`;
